File: benchmarks/comparable/list_ops.scala
// EXPECT: 9999900000 100000 50000
// list_ops.scala - build a 100000-element List, then map / filter / fold it.
// The ROADMAP's benchmark suite v1 "list-ops" workload (Phase 3). It prints the
// fold result, the built length and the filtered length, so a silently wrong
// collection surface fails the run instead of reading as a fast one.
// The three numbers were computed with tools/scala3-3.9.0, never by hand.

def build(n: Int): List[Int] =
  var xs = List[Int]()
  var i = n
  while i > 0 do
    i -= 1
    xs = i :: xs
  xs

@main def benchListOps(): Unit =
  val xs = build(100000)
  val doubled = xs.map(_ * 2)
  val evens = xs.filter(_ % 2 == 0)
  val total = doubled.foldLeft(0)(_ + _)
  println(total.toString + " " + xs.length + " " + evens.length)

File: benchmarks/comparable/python/list_ops.py
# list_ops.py - CPython/protoPython twin of benchmarks/comparable/list_ops.scala:
# build a 100000-element list, then map / filter / fold it.
# Result: 9999900000 100000 50000 (computed with tools/scala3-3.9.0).


def build(n):
    xs = []
    i = n
    while i > 0:
        i -= 1
        xs.insert(0, i)
    return xs


def main():
    xs = build(100000)
    doubled = [x * 2 for x in xs]
    evens = [x for x in xs if x % 2 == 0]
    total = 0
    for x in doubled:
        total += x
    print(str(total) + " " + str(len(xs)) + " " + str(len(evens)))


main()
